### PR TITLE
Support UniDiffuser model and pipeline

### PR DESCRIPTION
**Pull request by nemonameless**
### PR types
New features

### PR changes
Models

### Description

UniDiffuser ([One Transformer Fits All Distributions in Multi-Modal Diffusion](https://arxiv.org/abs/2303.06555))

## Inference

| Task                  |  Paddle Memory |  PyTorch  Memory |
| :-------------- | :---------------- | :----------------: |
|       Text-to-image Generation    | 7.01 GB |    8.65 GB      | 
|       Image-to-text Generation    | 5.88 GB |    7.62 GB      | 
|       Unconditional Joint Generation    | 7.10 GB |    8.66 GB  | 
|        Unconditional Image Generation | 7.02 GB |   8.65 GB   | 
|       Unconditional Text Generation    | 5.52 GB |    7.49 GB   | 
|       Image Variation    | 7.02 GB |    8.40 GB      | 
|       Text Variation    | 5.53 GB |    7.49 GB      | 

* **(1) Text-to-image Generation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/text_to_image_generation-unidiffuser.py
```

image

* **(2) Image-to-text Generation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/image_to_text_generation-unidiffuser.py
```

An image of an astronaut floating in space

* **(3) Unconditional Joint Generation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/unconditional_image_text_joint_generation-unidiffuser.py
```
image

A living room and dining area with a chandelier

* **(4) Unconditional Image Generation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/unconditional_image_generation-unidiffuser.py
```

image

* **(5) Unconditional Text Generation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/unconditional_text_generation-unidiffuser.py
```

A man standing in front of a bunch of a waterfall

* **(6) Image Variation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/image_variation-unidiffuser.py
```

image

* **(7) Text Variation**
```shell
CUDA_VISIBLE_DEVICES=0 python examples/inference/text_variation-unidiffuser.py
```

An elephant swimming in the water in the ocean

### CODE REVIEW

File: ppdiffusers/ppdiffusers/models/caption_decoder.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+from paddle import nn
+from paddle.nn import functional as F
+
+from paddlenlp.transformers import GPTConfig, GPTLMHeadModel, GPTTokenizer
+
+from ..configuration_utils import ConfigMixin, register_to_config
+from .modeling_utils import ModelMixin
+
+
+class CaptionDecoder(ModelMixin, ConfigMixin):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/caption_decoder.py]: 感觉这个 CaptionDecoder 可以以 paddlenlp的形式实现？这个需要讨论，可能后期还会同步HF

File: ppdiffusers/ppdiffusers/models/uvit.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+def timestep_embedding(timesteps, dim, max_period=10000):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/uvit.py]: 这个应该可以复用， https://github.com/PaddlePaddle/PaddleNLP/blob/706e5d6a41964106ca727dd5d3a4b09a823b376a/ppdiffusers/ppdiffusers/models/embeddings.py#L23

File: ppdiffusers/ppdiffusers/models/uvit.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+
+def interpolate_pos_emb(pos_emb, old_shape, new_shape):
+    pos_emb = einops.rearrange(pos_emb, "B (H W) C -> B C H W", H=old_shape[0], W=old_shape[1])
+    pos_emb = F.interpolate(pos_emb, new_shape, mode="bilinear")
+    pos_emb = einops.rearrange(pos_emb, "B C H W -> B (H W) C")
+    return pos_emb
+
+
+class Attention(nn.Layer):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/uvit.py:44]: 这应该也可以复用，https://github.com/PaddlePaddle/PaddleNLP/blob/706e5d6a41964106ca727dd5d3a4b09a823b376a/ppdiffusers/ppdiffusers/models/cross_attention.py#L27 只是权重命名和shape发生变化了

File: ppdiffusers/ppdiffusers/models/uvit.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+            attn = F.softmax(attn, axis=-1)
+            attn = self.attn_drop(attn)
+            x = paddle.mm(attn, v).transpose([0, 2, 1, 3]).reshape([B, L, C])
+
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class Block(nn.Layer):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/uvit.py:135]: https://github.com/PaddlePaddle/PaddleNLP/blob/706e5d6a41964106ca727dd5d3a4b09a823b376a/ppdiffusers/ppdiffusers/models/attention.py#L164 这个加个 DropPath 和 skip_linear 估计就能用了

File: ppdiffusers/ppdiffusers/models/uvit.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        x = x + self.drop_path(self.attn(x))
+        x = self.norm2(x)
+
+        x = x + self.drop_path(self.mlp(x))
+        x = self.norm3(x)
+
+        return x
+
+
+class PatchEmbed(nn.Layer):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/uvit.py]: https://github.com/PaddlePaddle/PaddleNLP/blob/706e5d6a41964106ca727dd5d3a4b09a823b376a/ppdiffusers/ppdiffusers/models/embeddings.py#L115 这个应该也可以复用，只要删除 pos_embed？

Comment (JunnYu) [reply ppdiffusers/ppdiffusers/models/uvit.py]: 他这里因为有很多token，因此最后在后面直接加了一个可以学习的pos embedding了

File: ppdiffusers/ppdiffusers/models/uvit.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        sample (`paddle.Tensor` of shape `(batch_size, num_channels, height, width)`):
+            Hidden states output. Output of last layer of model.
+    """
+
+    sample_img: paddle.Tensor
+    sample_clip_img: paddle.Tensor
+    sample_text: paddle.Tensor
+
+
+class UViTModel(ModelMixin, ConfigMixin):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/models/uvit.py:189]: 这里应该跟 https://github.com/PaddlePaddle/PaddleNLP/blob/706e5d6a41964106ca727dd5d3a4b09a823b376a/ppdiffusers/ppdiffusers/models/transformer_2d.py#L42 这个很像，不过这里面没有区分，in_blocks，mid_block，out_bloks，但是里面的模型结构都是 BasicTransformerBlock

File: ppdiffusers/examples/inference/unidiffuser.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+from paddlenlp.transformers import (
+    CLIPFeatureExtractor,
+    CLIPTextModel,
+    CLIPTokenizer,
+    CLIPVisionModelWithProjection,
+)
+from ppdiffusers import DPMSolverMultistepScheduler, UniDiffuserPipeline
+
+scheduler = DPMSolverMultistepScheduler()
```
Comment (JunnYu) [inline ppdiffusers/examples/inference/unidiffuser.py]: DPMSolverMultistepScheduler，ppdiffusers有实现，之后应该与pipeline在一起

File: ppdiffusers/ppdiffusers/pipelines/unidiffuser/pipeline_unidiffuser_i.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            clip_img,
+            text=text,
+            t_img=timesteps,
+            t_text=t_text,
+            data_type=paddle.zeros_like(t_text, dtype=paddle.int32) + data_type,
+        )
+        x_out = combine(z_out, clip_img_out)
+        return x_out
+
+    def sample_fn(self, z=None, clip_img=None, text=None):
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/pipelines/unidiffuser/pipeline_unidiffuser_i.py]: 之后这些全要使用 ppdiffusers里面的scheduler的调用形式。

File: ppdiffusers/ppdiffusers/models/caption_decoder.py (hunk 1 above)
Comment (JunnYu) [reply ppdiffusers/ppdiffusers/models/caption_decoder.py]: 感觉可以https://github.com/PaddlePaddle/PaddleNLP/blob/develop/ppdiffusers/ppdiffusers/pipelines/stable_diffusion_safe/safety_checker.py 仿照这里的模型写个paddlenlp版本的模型，放到这个位置来着

File: ppdiffusers/ppdiffusers/pipelines/unidiffuser/pipeline_unidiffuser_i2t.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import time
+from typing import Callable, List, Optional, Union
+
+import einops
```
Comment (JunnYu) [inline ppdiffusers/ppdiffusers/pipelines/unidiffuser/pipeline_unidiffuser_i2t.py]: 这个在我们的仓库里面，如果可能的话，尽可能不要使用

File: ppdiffusers/ppdiffusers/models/caption_decoder.py (hunk 1 above)
Comment (nemonameless) [reply ppdiffusers/ppdiffusers/models/caption_decoder.py]: done

File: ppdiffusers/ppdiffusers/models/uvit.py (hunk 2 above)
Comment (nemonameless) [reply ppdiffusers/ppdiffusers/models/uvit.py]: done

**Comment by JunnYu** [comment]
thu-ml/unidiffuser/text_encoder  和 thu-ml/unidiffuser/image_encoder 权重应该多传了，text里面多了vision的权重，vision里面多了text的权重，下载的时候可能多下载好多

**Comment by JunnYu** [comment]
- unidiffuser-t.txt
A WORRIED MACHINE SOUTH LAKE IN THE RAIN
- unidiffuser-i2t.txt
An image of an astronaut flying in space
- unidiffuser-joint.txt
A man embraces with rubble in the direction strike by Israeli air strikes during the Gaza Strip
- unidiffuser-t2i2t.txt
  Owns Owns Own Own Own Own Own Own Own Own Own Own Own Own Own Own Own Own Own Own Own Owns Own Own Owns Own Owns Own Owns Own Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns Owns
- unidiffuser-i2t2i.png
image
- unidiffuser-t2i.png
image
- unidiffuser-i.png
image
- unidiffuser-joint.png
image

File: ppdiffusers/examples/inference/unidiffuser.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+)
+from ppdiffusers import DPMSolverUniDiffuserScheduler, UniDiffuserPipeline
+from ppdiffusers.models import AutoencoderKL, UViTModel
+from ppdiffusers.pipelines.unidiffuser import CaptionDecoder
+from ppdiffusers.utils import load_image
+
+scheduler = DPMSolverUniDiffuserScheduler()
+generator = paddle.Generator().manual_seed(0)
+
+pipe = UniDiffuserPipeline(
```
Comment (JunnYu) [inline ppdiffusers/examples/inference/unidiffuser.py]: UniDiffuserPipeline直接from_pretrained更好

File: ppdiffusers/requirements.txt [hunk 11]
```diff
@@ -2,4 +2,5 @@ paddlenlp>=2.5.2
 safetensors
 ftfy
 regex
-Pillow
\ No newline at end of file
+Pillow
+git+https://github.com/arogozhnikov/einops.git
```
Comment (JunnYu) [inline ppdiffusers/requirements.txt]: 我python setup.py install 装不了

**Comment by JunnYu** [comment]
（1）image_to_text_generation-unidiffuser.py
astronaut in space in space spaceship spaceship space background
（2）image_variation-unidiffuser.py
image
（3） text_to_image_generation-unidiffuser.py
image
（4） text_variation-unidiffuser.py
An Elephant Underwater and Caring for Its World
（5）unconditional_image_generation-unidiffuser.py
image
（6）unconditional_image_text_joint_generation-unidiffuser.py
Used BMW X6 M Competition 30i
（7）unconditional_text_generation-unidiffuser.py
Outdoor Lighting For Unique Carpeted Garden Garden Design Ideas Carpeted Lights

**Review by JunnYu — APPROVED** [review]
LGTM

**Pull request merged**